core/utils/utils.go

```diff
@@ -866,65 +867,53 @@ const (
 
 // StartOnce sets the state to Started
 func (once *StartStopOnce) StartOnce(name string, fn func() error) error {
-	once.Lock()
-	defer once.Unlock()
+	success := once.state.CAS(int32(StartStopOnce_Unstarted), int32(StartStopOnce_Started))
```

Comment:
Could we just do something like:

```
func (once *StartStopOnce) StartOnce() {
ran := false
sync.Once(func() {
  ran = true
  once.started = true
}
if ran == false {
  return errors.New("already started")
}

})

Reply:
`once.started` would need to be an atomic in this case or it's unsafe to read from. `fn` also needs to be called while the lock is held, or we can call `StopOnce` before `StartOnce` finishes executing.